readme
The Following files will be there with the tar of the Driver:

	mcs9865.c
	mcs9865.h
	mcs9865-isa.c
	mcs9865-isa.h
	Makefile
	README	
	mcs9865

Building the Driver:
--------------------
  Change to the folder with the Starex source files. And run the following Command:
	$ make

  Once the driver is compliles the loadable module mcs9865.ko will be created in the same folder.

	**  '$' --this symbol represent the shell prompt on linux			

To build driver on kernels from v2.6.34 till latest:
----------------------------------------------------
  Since the device ID 9865 is included (as built-in) with kernel versions
    starts from v2.6.34 and above, the following procedure is required to
    follow to detect MCS9865 devices.

  a) Go to the path,
        "/usr/src/kernels/linux-x.y.z/drivers/serial"
        $ cd /usr/src/kernels/linux-x.y.z/drivers/serial/

  b) Open file "8250_pci.c".
    - find for the macro PCI_DEVICE_ID_NETMOS_9865.  It will be found in two places,
        1) In the struct definition "serial_pci_tbl[]"
          - Comment the below lines,

        "
                {       PCI_VENDOR_ID_NETMOS, PCI_DEVICE_ID_NETMOS_9865,
                        0xA000, 0x1000,
                        0, 0, pbn_b0_1_115200 },

                {       PCI_VENDOR_ID_NETMOS, PCI_DEVICE_ID_NETMOS_9865,
                        0xA000, 0x3004,
                        0, 0, pbn_b0_bt_4_115200 },
        "

        2) In the routine "pci_netmos_init"
          - No need do any action, as we already done (a).

Loading the Driver:
-------------------
	
  To load the driver use the following command:
	$ insmod mcs9865.ko
	$ insmod mcs9865-isa.ko

**  '$' --this symbol represent the shell prompt on linux			

Unloading the Driver:
---------------------

  To unload the driver use the following command:
	$rmmod mcs9865
	$rmmod mcs9865-isa

**  '$' --this symbol represent the shell prompt on linux			

Installing the Driver:
----------------------
  To install the driver use the following command:
		
	make install

Uninstalling the Driver:
------------------------

  To un-install the driver use the following command:
		
	make uninstall

Note: only baud rates upto 3.0Mbps are implemented in the current driver.
	
Feature Configuration:
----------------------

  This driver comes with a static array uart_9865_contxts[ ] for all the 16 ports features in the 9865.c file.
  Where u can set the features like:
	-DMA in RX
	-DMA in TX
	-enable/disable Flow control
	-Flow control type
	-RX fifo trigger level
	-TX fifo trigger level
	-x-on character
	-x-off character

  An example array setting for PORT 0 which will be recognised as /dev/ttyD0:			

  //Port 0	--**This represents that this setting is for port 0 which will be recognised as /dev/ttyD0
  {
	.rx_dma_en	= 0,
	.tx_dma_en	= 0,
	.uart_mode	= MCS9865_RS232_MODE,
	.en_flow_control	= 0,
	.flow_ctrl_type 	= MCS9865_XON_XOFF_HW_FLOWCONTROL,
	.rxfifotrigger	= 64,
	.txfifotrigger	= 64,		
	.x_on		= SERIAL_DEF_XON,
	.x_off		= SERIAL_DEF_XOFF,
  },

	To change it to different modes like RS232 and RS422, we have to fille the .uart_mode with corresponding //UART mode defined in header files. for example for setting it to RS422 mode, we have to fill .uart_mode=MCS9865_RS422_MODE.


Kernels:
--------
  This driver is currently developed and tested on 2.6.11 linux kernel and above 	

Limitations:
------------
  This driver supports a maximum of 16 serial devices. This can be modified as required		

Steps for setting parallel port :	
---------------------------------
  1. rmmod lp
  2. rmmod parport_pc
  2. insmod /usr/lib/modules/2.6.15/kernel/drivers/parport/parport_pc io=0xb000 irq=10.

Note: Here, the io and irq should be noted from lspci -v.

Steps for setting serial isa combinatios:
-----------------------------------------
  1.Setting ISA 4 serial combination:
  ------------------------------------
	1.Go to mcs9865-isa.c file and go to the functon "struct pci_device_id".
 	2.Make comments to all other ISA combinations other than ISA 4 serial.

similarly for other ISA combinations.

PCI_SUBVEN_ID_9865_4S represents ISA 4 serial.
PCI_SUBVEN_ID_9865_3S represents ISA 3 serial.
PCI_SUBVEN_ID_9865_2S represents ISA 2 serial.
PCI_SUBVEN_ID_9865_1S represents ISA 1 serial.
PCI_SUBVEN_ID_9865_2P represents ISA 2 parallel.
PCI_SUBVEN_ID_9865_1S1P represents ISA 1 serial 1 parallel.
PCI_SUBVEN_ID_9865_2S1P represents ISA 2 serial 1 parallel.
		 


ReleaseNotes
   				Release notes for MCS9865 serial
Version 1.0.0.12
----------------
10-03-2011
1) Makefile is modified in giving the softlink name for the script 'mcs9865'.
   Followed as the Debian/Fedora policy.
2) No change in driver source.

Version 1.0.0.11
----------------
23-02-2011
1) Makefile is modified.
   - Added conditional check to support debian distribution, in creating
     soft symbolic links to /etc/rcS.d/.
   - This also works for fedora as usual.
2) No change in driver source.
3) Base version taken Version 1.0.0.10.
    
Version 1.0.0.10
----------------
17-12-10

1) Added support upto Linux kernel 2.6.36
2) No change in driver source.
3) Modified readme file: given user information to support driver on kernels
   from v2.6.34 to v2.6.36.

Version 1.0.0.9
---------------
18-05-10

Added support upto Linux kernel 2.6.33.

Version 1.0.0.8
---------------
Backward compatible with linux-2.6.18 for the support setserial application.

Version 1.0.0.7
---------------
Bug resolved for setserial application when using autoconfig option.

Version 1.0.0.6
----------------
Date 12-10-2009

Added support upto 2.6.31 kernels.

Version 1.0.0.5
---------------
Date 18-03-2009

Added backward compatibility support for 2.6.14 kernel.

Version 1.0.0.4
---------------
Date 21-01-2009

Added support for 2.6.24 and 2.6.25 kernels.

Version 1.0.0.3
-------------
Date 08-01-2009

Bugs fixed:
Added support for different ISA combinations.

Version 1.0.0.2
-------------
Date 6-11-2008

Bugs fixed:
Fixed Baud rate setting issue when changed from higher to lower baudrate



Version 1.0.0.0
-------------
Date 18-4-2008

1. Initial version

Tested 
1. Differrent BAUD rates, file transfers.





mcs9865.h
/*
 *  linux/drivers/serial/9865.h
 *
 *  Based on drivers/serial/8250.c by Russell King.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This code is modified to support moschip 9865 series serial devices
 */

#if LINUX_VERSION_CODE <= KERNEL_VERSION(2,6,15)
#include <linux/config.h>
#endif


struct old_serial_port {
	unsigned int 	uart;
	unsigned int 	baud_base;
	unsigned int	 port;
	unsigned int	 irq;
	unsigned int 	flags;
	unsigned char 	hub6;
	unsigned char 	io_type;
	unsigned char 	*iomem_base;
	unsigned short 	iomem_reg_shift;
};

struct serial9865_config {
	unsigned short	fifo_size;
	unsigned short	tx_loadsz;
	unsigned char		fcr;
	unsigned int		flags;
};

#define DIV 1

#define UART_CAP_FIFO	(1 << 8)	/* UART has FIFO */
#define UART_CAP_EFR	(1 << 9)	/* UART has EFR */
#define UART_CAP_SLEEP	(1 << 10)	/* UART has IER sleep */
#define UART_CAP_AFE	(1 << 11)	/* MCR-based hw flow control */
#define UART_CAP_UUE	(1 << 12)	/* UART needs IER bit 6 set (Xscale) */

#define REG_TX_DMA_START_ADDRESS_LOW	((0x80)/DIV)
#define REG_TX_DMA_START_ADDRESS_HIGH	((0x84)/DIV)
#define REG_TX_DMA_LENGTH				((0x88)/DIV)
#define REG_TX_DMA_START				((0x8C)/DIV)
#define REG_TX_DMA_STOP					((0x90)/DIV)
#define REG_TX_DMA_STOP_DONE			((0x94)/DIV)
#define REG_TX_BYTES_TRANSFERRED		((0x98)/DIV)
#define REG_TX_DMA_BUSY					((0x9C)/DIV)
#define REG_TX_DMA_DONE					((0xA0)/DIV)
#define REG_TX_RDY_1					((0xA4)/DIV)

#define REG_RX_DMA_START_ADDRESS_LOW	((0x100)/DIV)
#define REG_RX_DMA_START_ADDRESS_HIGH	((0x104)/DIV)
#define REG_RX_DMA_LENGTH				((0x108)/DIV)
#define REG_RX_DMA_START				((0x10C)/DIV)
#define REG_RX_DMA_STOP					((0x110)/DIV)
#define REG_RX_TRIG_LVL					((0x114)/DIV)
#define REG_RX_DMA_STOP_DONE			((0x118)/DIV)
#define REG_RX_BYTES_NEED_TO_RECV		((0x11C)/DIV)
#define REG_RX_DMA_BUSY					((0x120)/DIV)
#define REG_RX_DMA_DONE					((0x124)/DIV)
#define REG_RX_RDY_1					((0x128)/DIV)
#define REG_RX_MEM_4K_LMT				((0x12C)/DIV)


#define REG_GLBL_ISR			((0x3A0)/DIV)
#define REG_GLBL_ICLR			((0x3A4)/DIV)
#define REG_GLBL_IER			((0x3A8)/DIV)

#define TX_DMA_START_BIT		1<<0
#define TX_DMA_STOP_BIT			1<<0
#define TX_DMA_STOP_DONE_BIT 	1<<0
#define TX_DMA_DONE				1<<0
#define TX_DMA_BUSY				1<<0
#define TX_DMA_RDY				1<<0

#define RX_DMA_START_BIT		1<<0
#define RX_DMA_STOP_BIT			1<<0
#define RX_DMA_STOP_DONE_BIT 	1<<0
#define RX_DMA_DONE				1<<0
#define RX_DMA_BUSY				1<<0
#define RX_DMA_RDY				1<<0

#define SPINTR_DMA					0x01
#define SPINTR_TXDMA_ABORT_DONE		0x02
#define SPINTR_TXDMA_STOP_DONE		0x04
#define SPINTR_TXDMA_DONE 			0x08
#define SPINTR_RXDMA_ABORT_DONE 	0x10
#define SPINTR_RXDMA_STOP_DONE		0x20
#define SPINTR_RXDMA_DONE			0x40
#define SPINTR_RXDMA_PARTDONE		0x80

#define SERIAL_450MODE		0x5470
#define SERIAL_550MODE		0x5471
#define SERIAL_550AMODE		0x5472
#define SERIAL_650MODE		0x5473
#define SERIAL_750MODE		0x5474
#define SERIAL_850MODE		0x5475
#define SERIAL_950MODE		0x5476

// Default xon/xoff characters.
#define SERIAL_DEF_XON		0x11
#define SERIAL_DEF_XOFF		0x13

// UART mode
#define MCS9865_RS232_MODE					0
#define MCS9865_RS422_MODE					1
#define MCS9865_RS485_HALF_DUPLEX			2
#define MCS9865_RS485_HALF_DUPLEX_ECHO		4
#define MCS9865_RS485_FULL_DUPLEX			5
#define MCS9865_DTR_DSR_HW_FLOWCONTROL 		6
#define MCS9865_XON_XOFF_HW_FLOWCONTROL		7
#define MCS9865_RTS_CTS_HW_FLOWCONTROL		8
#define MCS9865_IRDA_MODE					9

#define PORT_ENHANC			14

//CommSet Registers
//Common Registers Set (memory mapped)
#define SER_DCR_DIN_REG			((0x200)/DIV)
#define SER_VEN_REG				((0x204)/DIV)
#define SP_SIGNAL_VALID_REG		((0x208)/DIV)
#define SP_IO_CONTROL_ENABLE_REG	((0x20C)/DIV)
#define SP_OUTPUT_REG_VALID		((0x210)/DIV)
#define SP_TX_TRIGGER_LVL		((0x21C)/DIV)
#define SP_CLK_SELECT_REG		((0x214)/DIV)
#define PP_DIV_REG				((0x250)/DIV)
#define PP_RX_TRIG_LEVEL		((0x254)/DIV)
#define PP_TX_TRIG_LEVEL		((0x258)/DIV)
#define PP_PERI_HOST_HIGH_REG	((0x25C)/DIV)
#define SER_TFL_REG				((0x220)/DIV) 
#define SER_RFL_REG				((0x224)/DIV) 
#define SER_SOFT_RESET_REG		((0x238)/DIV) 

#define COM_DMA_MODE_EN			0x10000000
#define COM_550EX_MODE_EN		0x00001000


#define SER_TFL_REG				((0x220)/DIV)
#define SER_RFL_REG				((0x224)/DIV)
#define SER_SOFT_RESET_REG		((0x238)/DIV)


#define PCI_DEVICE_ID_NETMOS_9865 0x9865
#define PCI_SUBVEN_ID_9865	  0x1000
#define PCI_SUBDEV_ID_9865	  0xa000	

#define BAR_IO		0x001
#define BAR_MEM 	0x000
#define BAR_FMT  	0x001

#define DMA_RX_SZ 		256
#define DMA_RX_BUFFER_SZ 	131072
#define DMA_TX_BUFFER_SZ 	4096

#if defined(__i386__) && (defined(CONFIG_M386) || defined(CONFIG_M486))
#define _INLINE_ inline
#else
#define _INLINE_
#endif

#define DEFAULT9865_BAUD 115200

static unsigned int serial9865_get_mctrl(struct uart_port *);


mcs9865-isa.h
/*
 *  linux/drivers/serial/9865-isa.h
 *
 *  Driver for PCI based ISA 8250/16550-type serial ports
 *
 *  Based on drivers/serial/8250.h by Russell King
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 */
#include <linux/version.h>

#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,18)
#include <linux/config.h>
#endif

#define MAX_ISA_PORTS 32
#define STX_NR_PORTS 0x000F

#define UART_CAP_FIFO	(1 << 8)	/* UART has FIFO */
#define UART_CAP_EFR	(1 << 9)	/* UART has EFR */
#define UART_CAP_SLEEP	(1 << 10)	/* UART has IER sleep */
#define UART_CAP_AFE	(1 << 11)	/* MCR-based hw flow control */
#define UART_CAP_UUE	(1 << 12)	/* UART needs IER bit 6 set (Xscale) */

#define BAR_IO 	0x001
#define BAR_MEM 0x000
#define BAR_FMT 0x001

#define DEFAULT9865_BAUD 115200

#define PCI_DEVICE_ID_NETMOS_9865 0x9865
#define PCI_SUBDEV_ID_9865      0xa000

#define PCI_SUBVEN_ID_9865_4S      0x3004
#define PCI_SUBVEN_ID_9865_3S      0x3003
#define PCI_SUBVEN_ID_9865_2S      0x3002
#define PCI_SUBVEN_ID_9865_1S      0x3001
#define PCI_SUBVEN_ID_9865_2P      0x3020
#define PCI_SUBVEN_ID_9865_1S1P      0x3011
#define PCI_SUBVEN_ID_9865_2S1P      0x3012


mcs9865-isa.c
/*
 *  linux/drivers/serial/9865-isa.c
 *
 *  Probe module for 9865  PCI based ISA serial ports.
 *
 *  Based on drivers/char/8250_pci.c, by Russell King.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License.
 *
 */
#include <linux/module.h>
#include <linux/init.h>
#include <linux/pci.h>
#include <linux/sched.h>
#include <linux/string.h>
#include <linux/kernel.h>
#include <linux/slab.h>
#include <linux/delay.h>
#include <linux/tty.h>
#include <linux/serial_core.h>
#include <linux/8250_pci.h>
#include <linux/bitops.h>
#include <linux/serial_8250.h>

#include <asm/byteorder.h>
#include <asm/io.h>

#include "mcs9865-isa.h"

#undef SERIAL_DEBUG_PCI

/*
 * Definitions for PCI support.
 */
#define FL_BASE_MASK	0x0007
#define FL_BASE0		0x0000
#define FL_BASE1		0x0001
#define FL_BASE2		0x0002
#define FL_BASE3		0x0003
#define FL_BASE4		0x0004
#define FL_BASE5		0x0005
#define FL_GET_BASE(x)	(x & FL_BASE_MASK)

/* Use successive BARs (PCI base address registers),
   else use offset into some specified BAR */
#define FL_BASE_BARS		0x0008

#define PCI_NUM_BAR_RESOURCES	6

struct isa_private {
		int nr_ports;
		int line[MAX_ISA_PORTS];	
}isa_priv;

#if 0
#define DEBUG(fmt...)   printk(fmt)
#else
#define DEBUG(fmt...)   do { } while (0)
#endif

/*
 * Probe one serial board.  Unfortunately, there is no rhyme nor reason
 * to the arrangement of serial ports on a PCI card.
 */
static int __devinit
pci_isa_serial_init_one(struct pci_dev *dev, const struct pci_device_id *ent)
{
	int rc,retval,i,nr_ports;
	rc = pci_enable_device(dev);
	if (rc)
		return rc;

        //To verify wether it is a serial communication hardware
        if ((dev->class >> 8) !=  PCI_CLASS_COMMUNICATION_OTHER){
                DEBUG("Not a communication hardware\n");
                retval = -ENODEV;
                goto disable;
        }
	
	//To verify wether it is a MCS9865 ISA type BARs
        if(((pci_resource_flags(dev,FL_BASE0) & BAR_FMT) ^ BAR_IO) || ((pci_resource_flags(dev,FL_BASE1) & BAR_FMT) ^ BAR_IO) ||((pci_resource_flags(dev,FL_BASE2) & BAR_FMT) ^ BAR_IO) || ((pci_resource_flags(dev,FL_BASE3) & BAR_FMT) ^ BAR_IO) || ((pci_resource_flags(dev,FL_BASE5) & BAR_FMT) ^ BAR_MEM)) {
                DEBUG("Not a MCS9865 ISA type device\n");
                retval = -ENOMEM;
                goto disable;
        }

	pci_set_master(dev);
	nr_ports=dev->subsystem_device & STX_NR_PORTS;

	DEBUG("No of ports detected is =%d \n",nr_ports);	

	for (i = 0; i < nr_ports; i++) {
		struct uart_port serial_port;
		memset(&serial_port, 0, sizeof(struct uart_port));

		serial_port.flags = UPF_SKIP_TEST | UPF_SHARE_IRQ;
		serial_port.uartclk = DEFAULT9865_BAUD * 16;
		serial_port.irq = dev->irq;
		serial_port.dev = &dev->dev;
		serial_port.iobase = pci_resource_start(dev,i);
		serial_port.iotype = UPIO_PORT;

		retval = serial8250_register_port(&serial_port);
		if (retval < 0) {
			printk(KERN_WARNING "Couldn't register serial port %s: %d\n", pci_name(dev), retval);
			goto disable;
		}

		isa_priv.line[isa_priv.nr_ports]=retval;
		isa_priv.nr_ports++;

	}

	return 0;

 disable:
	pci_disable_device(dev);
	return retval;
}

static void __devexit pci_isa_serial_remove_one(struct pci_dev *dev)
{
	int i;
	for (i = 0; i < isa_priv.nr_ports; i++)
		serial8250_unregister_port(isa_priv.line[i]);
}

static struct pci_device_id serial_pci_isa_tbl[] = {
	//{PCI_VENDOR_ID_NETMOS, PCI_DEVICE_ID_NETMOS_9865, PCI_SUBVEN_ID_9865,PCI_SUBDEV_ID_9865,0,0,0},
	{PCI_VENDOR_ID_NETMOS, PCI_DEVICE_ID_NETMOS_9865, PCI_SUBDEV_ID_9865, PCI_SUBVEN_ID_9865_4S, 0,0,0},
	{PCI_VENDOR_ID_NETMOS, PCI_DEVICE_ID_NETMOS_9865, PCI_SUBDEV_ID_9865, PCI_SUBVEN_ID_9865_3S, 0,0,0},
	{PCI_VENDOR_ID_NETMOS, PCI_DEVICE_ID_NETMOS_9865, PCI_SUBDEV_ID_9865, PCI_SUBVEN_ID_9865_2S, 0,0,0},
	{PCI_VENDOR_ID_NETMOS, PCI_DEVICE_ID_NETMOS_9865, PCI_SUBDEV_ID_9865, PCI_SUBVEN_ID_9865_1S, 0,0,0},
	{PCI_VENDOR_ID_NETMOS, PCI_DEVICE_ID_NETMOS_9865, PCI_SUBDEV_ID_9865, PCI_SUBVEN_ID_9865_2P, 0,0,0},
	{PCI_VENDOR_ID_NETMOS, PCI_DEVICE_ID_NETMOS_9865, PCI_SUBDEV_ID_9865, PCI_SUBVEN_ID_9865_1S1P, 0,0,0},
	{PCI_VENDOR_ID_NETMOS, PCI_DEVICE_ID_NETMOS_9865, PCI_SUBDEV_ID_9865, PCI_SUBVEN_ID_9865_2S1P, 0,0,0},
	{ 0, }
};

static struct pci_driver serial_pci_isa_driver = {
	.name		= "serial-isa",
	.probe		= pci_isa_serial_init_one,
	.remove		= __devexit_p(pci_isa_serial_remove_one),
	.id_table	= serial_pci_isa_tbl,
};

static int __init serial_isa_init(void)
{	
	isa_priv.nr_ports=0;
	return pci_register_driver(&serial_pci_isa_driver);
}

static void __exit serial_isa_exit(void)
{
	pci_unregister_driver(&serial_pci_isa_driver);
}

module_init(serial_isa_init);
module_exit(serial_isa_exit);

MODULE_LICENSE("GPL");
MODULE_DESCRIPTION("MCS9865 ISA serial probe module");
MODULE_DEVICE_TABLE(pci, serial_pci_isa_tbl);
